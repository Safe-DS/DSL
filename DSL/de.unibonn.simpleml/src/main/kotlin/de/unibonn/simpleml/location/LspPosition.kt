package de.unibonn.simpleml.location

/**
 * A specific position in a program using the zero-based indexing of LSP.
 */
data class LspPosition(val line: LspLine, val column: LspColumn) : Comparable<LspPosition> {
    companion object {

        @JvmStatic
        fun fromInts(line: Int, column: Int): LspPosition {
            return LspPosition(
                LspLine(line),
                LspColumn(column)
            )
        }
    }

    fun toXtextPosition(): XtextPosition {
        return XtextPosition(
            line.toXtextLine(),
            column.toXtextColumn()
        )
    }

    override fun toString(): String {
        return "[$line, $column]"
    }

    override operator fun compareTo(other: LspPosition): Int {
        val lineComparison = this.line.compareTo(other.line)
        if (lineComparison != 0) {
            return lineComparison
        }

        return this.column.compareTo(other.column)
    }
}

/**
 * A line in a program. Counting starts at 0.
 *
 * @throws IllegalArgumentException If value is negative.
 */
@JvmInline
value class LspLine(val value: Int) : Comparable<LspLine> {
    init {
        require(value >= 0) { "Line must be at least 0." }
    }

    fun toXtextLine(): XtextLine {
        return XtextLine(value + 1)
    }

    override fun toString(): String {
        return value.toString()
    }

    override operator fun compareTo(other: LspLine): Int {
        return this.value.compareTo(other.value)
    }
}

/**
 * A column in a program. Counting starts at 0.
 *
 * @throws IllegalArgumentException If value is negative.
 */
@JvmInline
value class LspColumn(val value: Int) : Comparable<LspColumn> {
    init {
        require(value >= 0) { "Column must be at least 0." }
    }

    fun toXtextColumn(): XtextColumn {
        return XtextColumn(value + 1)
    }

    override fun toString(): String {
        return value.toString()
    }

    override operator fun compareTo(other: LspColumn): Int {
        return this.value.compareTo(other.value)
    }
}
