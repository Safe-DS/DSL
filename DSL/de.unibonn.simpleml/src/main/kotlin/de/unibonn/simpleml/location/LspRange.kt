package de.unibonn.simpleml.location

/**
 * A range in a program from a start to an end position with some length using the zero-based indexing of LSP.
 *
 * @see LspPosition
 * @see ProgramRangeLength
 */
data class LspRange(val start: LspPosition, val end: LspPosition, val length: ProgramRangeLength) {

    companion object {

        @JvmStatic
        fun fromInts(startLine: Int, startColumn: Int, endLine: Int, endColumn: Int, length: Int): LspRange {
            return LspRange(
                LspPosition.fromInts(startLine, startColumn),
                LspPosition.fromInts(endLine, endColumn),
                ProgramRangeLength(length)
            )
        }
    }

    fun toXtextRange(): XtextRange {
        return XtextRange(
            start.toXtextPosition(),
            end.toXtextPosition(),
            length
        )
    }

    override fun toString(): String {
        return "[$start .. $end]"
    }
}
