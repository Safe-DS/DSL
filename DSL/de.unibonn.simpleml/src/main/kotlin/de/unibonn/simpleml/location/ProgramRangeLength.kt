package de.unibonn.simpleml.location

/**
 * The number of characters in a program range (Xtext/LSP). This value must be non-negative.
 *
 * @throws IllegalArgumentException If value is negative.
 *
 * @see XtextRange
 * @see LspRange
 */
@JvmInline
value class ProgramRangeLength(val value: Int) {
    init {
        require(value >= 0) { "Length must be at least 0." }
    }

    override fun toString(): String {
        val chars = if (value == 1) "char" else "chars"
        return "$value $chars"
    }
}
