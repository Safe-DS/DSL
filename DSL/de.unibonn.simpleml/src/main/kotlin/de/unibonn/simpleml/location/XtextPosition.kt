package de.unibonn.simpleml.location

/**
 * A specific position in a program using the one-based indexing of Xtext.
 */
data class XtextPosition(val line: XtextLine, val column: XtextColumn) : Comparable<XtextPosition> {
    companion object {

        @JvmStatic
        fun fromInts(line: Int, column: Int): XtextPosition {
            return XtextPosition(
                XtextLine(line),
                XtextColumn(column)
            )
        }
    }

    fun toLspPosition(): LspPosition {
        return LspPosition(
            line.toLspLine(),
            column.toLspColumn()
        )
    }

    override fun toString(): String {
        return "$line:$column"
    }

    override operator fun compareTo(other: XtextPosition): Int {
        val lineComparison = this.line.compareTo(other.line)
        if (lineComparison != 0) {
            return lineComparison
        }

        return this.column.compareTo(other.column)
    }
}

/**
 * A line in a program. Counting starts at 1.
 *
 * @throws IllegalArgumentException If value is less than 1.
 */
@JvmInline
value class XtextLine(val value: Int) : Comparable<XtextLine> {
    init {
        require(value >= 1) { "Line must be at least 1." }
    }

    fun toLspLine(): LspLine {
        return LspLine(value - 1)
    }

    override fun toString(): String {
        return value.toString()
    }

    override operator fun compareTo(other: XtextLine): Int {
        return this.value.compareTo(other.value)
    }
}

/**
 * A column in a program. Counting starts at 1.
 *
 * @throws IllegalArgumentException If value is less than 1.
 */
@JvmInline
value class XtextColumn(val value: Int) : Comparable<XtextColumn> {
    init {
        require(value >= 1) { "Column must be at least 1." }
    }

    fun toLspColumn(): LspColumn {
        return LspColumn(value - 1)
    }

    override fun toString(): String {
        return value.toString()
    }

    override operator fun compareTo(other: XtextColumn): Int {
        return this.value.compareTo(other.value)
    }
}
