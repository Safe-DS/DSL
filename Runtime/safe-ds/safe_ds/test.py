import numpy as np
import pandas as pd

from safe_ds.data import Table, TableSchema, Row

table = Table.from_json("../tests/resources/test_schema_table.json")
rows_is: list[Row] = table.to_rows()
table_is: Table = Table.from_rows(rows_is)

print(table._data)
print("\n")
print(table_is._data)
print("\n")
for (index, series_row) in table._data.iterrows():
    print(series_row)
print("\n")
for row in rows_is:
    print(row._data)
