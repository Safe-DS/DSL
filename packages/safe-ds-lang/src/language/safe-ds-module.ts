import { DeepPartial, inject, Module } from 'langium';
import {
    createDefaultModule,
    createDefaultSharedModule,
    DefaultSharedModuleContext,
    LangiumServices,
    LangiumSharedServices,
    PartialLangiumServices,
} from 'langium/lsp';
import { SafeDsAnnotations } from './builtins/safe-ds-annotations.js';
import { SafeDsClasses } from './builtins/safe-ds-classes.js';
import { SafeDsEnums, SafeDsImpurityReasons } from './builtins/safe-ds-enums.js';
import { SafeDsCommentProvider } from './documentation/safe-ds-comment-provider.js';
import { SafeDsDocumentationProvider } from './documentation/safe-ds-documentation-provider.js';
import { SafeDsCallGraphComputer } from './flow/safe-ds-call-graph-computer.js';
import { SafeDsGeneratedModule, SafeDsGeneratedSharedModule } from './generated/module.js';
import { SafeDsPythonGenerator } from './generation/safe-ds-python-generator.js';
import { SafeDsValueConverter } from './grammar/safe-ds-value-converter.js';
import { SafeDsNodeMapper } from './helpers/safe-ds-node-mapper.js';
import { SafeDsCallHierarchyProvider } from './lsp/safe-ds-call-hierarchy-provider.js';
import { SafeDsDocumentSymbolProvider } from './lsp/safe-ds-document-symbol-provider.js';
import { SafeDsFormatter } from './lsp/safe-ds-formatter.js';
import { SafeDsInlayHintProvider } from './lsp/safe-ds-inlay-hint-provider.js';
import { SafeDsNodeInfoProvider } from './lsp/safe-ds-node-info-provider.js';
import { SafeDsNodeKindProvider } from './lsp/safe-ds-node-kind-provider.js';
import { SafeDsSemanticTokenProvider } from './lsp/safe-ds-semantic-token-provider.js';
import { SafeDsSignatureHelpProvider } from './lsp/safe-ds-signature-help-provider.js';
import { SafeDsTypeHierarchyProvider } from './lsp/safe-ds-type-hierarchy-provider.js';
import { SafeDsPartialEvaluator } from './partialEvaluation/safe-ds-partial-evaluator.js';
import { SafeDsScopeComputation } from './scoping/safe-ds-scope-computation.js';
import { SafeDsScopeProvider } from './scoping/safe-ds-scope-provider.js';
import { SafeDsClassHierarchy } from './typing/safe-ds-class-hierarchy.js';
import { SafeDsCoreTypes } from './typing/safe-ds-core-types.js';
import { SafeDsTypeChecker } from './typing/safe-ds-type-checker.js';
import { SafeDsTypeComputer } from './typing/safe-ds-type-computer.js';
import { registerValidationChecks } from './validation/safe-ds-validator.js';
import { SafeDsDocumentBuilder } from './workspace/safe-ds-document-builder.js';
import { SafeDsPackageManager } from './workspace/safe-ds-package-manager.js';
import { SafeDsWorkspaceManager } from './workspace/safe-ds-workspace-manager.js';
import { SafeDsPurityComputer } from './purity/safe-ds-purity-computer.js';
import { SafeDsSettingsProvider } from './workspace/safe-ds-settings-provider.js';
import { SafeDsRenameProvider } from './lsp/safe-ds-rename-provider.js';
import { SafeDsRunner } from './runner/safe-ds-runner.js';
import { SafeDsTypeFactory } from './typing/safe-ds-type-factory.js';

/**
 * Declaration of custom services - add your own service classes here.
 */
export type SafeDsAddedServices = {
    builtins: {
        Annotations: SafeDsAnnotations;
        Classes: SafeDsClasses;
        Enums: SafeDsEnums;
        ImpurityReasons: SafeDsImpurityReasons;
    };
    evaluation: {
        PartialEvaluator: SafeDsPartialEvaluator;
    };
    flow: {
        CallGraphComputer: SafeDsCallGraphComputer;
    };
    generation: {
        PythonGenerator: SafeDsPythonGenerator;
    };
    helpers: {
        NodeMapper: SafeDsNodeMapper;
    };
    lsp: {
        NodeInfoProvider: SafeDsNodeInfoProvider;
    };
    purity: {
        PurityComputer: SafeDsPurityComputer;
    };
    types: {
        ClassHierarchy: SafeDsClassHierarchy;
        CoreTypes: SafeDsCoreTypes;
        TypeChecker: SafeDsTypeChecker;
        TypeComputer: SafeDsTypeComputer;
        TypeFactory: SafeDsTypeFactory;
    };
    workspace: {
        PackageManager: SafeDsPackageManager;
        SettingsProvider: SafeDsSettingsProvider;
    };
    runtime: {
        Runner: SafeDsRunner;
    };
};

/**
 * Union of Langium default services and your custom services - use this as constructor parameter
 * of custom service classes.
 */
export type SafeDsServices = LangiumServices & SafeDsAddedServices;

/**
 * Dependency injection module that overrides Langium default services and contributes the
 * declared custom services. The Langium defaults can be partially specified to override only
 * selected services, while the custom services must be fully specified.
 */
export const SafeDsModule: Module<SafeDsServices, PartialLangiumServices & SafeDsAddedServices> = {
    builtins: {
        Annotations: (services) => new SafeDsAnnotations(services),
        Classes: (services) => new SafeDsClasses(services),
        Enums: (services) => new SafeDsEnums(services),
        ImpurityReasons: (services) => new SafeDsImpurityReasons(services),
    },
    documentation: {
        CommentProvider: (services) => new SafeDsCommentProvider(services),
        DocumentationProvider: (services) => new SafeDsDocumentationProvider(services),
    },
    evaluation: {
        PartialEvaluator: (services) => new SafeDsPartialEvaluator(services),
    },
    flow: {
        CallGraphComputer: (services) => new SafeDsCallGraphComputer(services),
    },
    generation: {
        PythonGenerator: (services) => new SafeDsPythonGenerator(services),
    },
    helpers: {
        NodeMapper: (services) => new SafeDsNodeMapper(services),
    },
    lsp: {
        CallHierarchyProvider: (services) => new SafeDsCallHierarchyProvider(services),
        DocumentSymbolProvider: (services) => new SafeDsDocumentSymbolProvider(services),
        Formatter: () => new SafeDsFormatter(),
        InlayHintProvider: (services) => new SafeDsInlayHintProvider(services),
        NodeInfoProvider: (services) => new SafeDsNodeInfoProvider(services),
        RenameProvider: (services) => new SafeDsRenameProvider(services),
        SemanticTokenProvider: (services) => new SafeDsSemanticTokenProvider(services),
        SignatureHelp: (services) => new SafeDsSignatureHelpProvider(services),
        TypeHierarchyProvider: (services) => new SafeDsTypeHierarchyProvider(services),
    },
    parser: {
        ValueConverter: () => new SafeDsValueConverter(),
    },
    purity: {
        PurityComputer: (services) => new SafeDsPurityComputer(services),
    },
    references: {
        ScopeComputation: (services) => new SafeDsScopeComputation(services),
        ScopeProvider: (services) => new SafeDsScopeProvider(services),
    },
    types: {
        ClassHierarchy: (services) => new SafeDsClassHierarchy(services),
        CoreTypes: (services) => new SafeDsCoreTypes(services),
        TypeChecker: (services) => new SafeDsTypeChecker(services),
        TypeComputer: (services) => new SafeDsTypeComputer(services),
        TypeFactory: (services) => new SafeDsTypeFactory(services),
    },
    workspace: {
        PackageManager: (services) => new SafeDsPackageManager(services),
        SettingsProvider: (services) => new SafeDsSettingsProvider(services),
    },
    runtime: {
        Runner: (services) => new SafeDsRunner(services),
    },
};

export type SafeDsSharedServices = LangiumSharedServices;

export const SafeDsSharedModule: Module<SafeDsSharedServices, DeepPartial<SafeDsSharedServices>> = {
    lsp: {
        NodeKindProvider: () => new SafeDsNodeKindProvider(),
    },
    workspace: {
        DocumentBuilder: (services) => new SafeDsDocumentBuilder(services),
        WorkspaceManager: (services) => new SafeDsWorkspaceManager(services),
    },
};

/**
 * Create the full set of services required by Langium.
 *
 * First inject the shared services by merging two modules:
 *  - Langium default shared services
 *  - Services generated by langium-cli
 *
 * Then inject the language-specific services by merging three modules:
 *  - Langium default language-specific services
 *  - Services generated by langium-cli
 *  - Services specified in this file
 *
 * @param context Optional module context with the LSP connection.
 * @param options Further options to configure the Safe-DS module.
 * @return An object wrapping the shared services and the language-specific services.
 */
export const createSafeDsServices = function (
    context: DefaultSharedModuleContext,
    options?: ModuleOptions,
): {
    shared: LangiumSharedServices;
    SafeDs: SafeDsServices;
} {
    const shared = inject(createDefaultSharedModule(context), SafeDsGeneratedSharedModule, SafeDsSharedModule);
    const SafeDs = inject(createDefaultModule({ shared }), SafeDsGeneratedModule, SafeDsModule);
    shared.ServiceRegistry.register(SafeDs);
    registerValidationChecks(SafeDs);
    SafeDs.runtime.Runner.updateRunnerCommand(options?.runnerCommand);
    return { shared, SafeDs };
};

/**
 * Create the full set of services required by Langium and load all builtin files into the workspace.
 *
 * First inject the shared services by merging two modules:
 *  - Langium default shared services
 *  - Services generated by langium-cli
 *
 * Then inject the language-specific services by merging three modules:
 *  - Langium default language-specific services
 *  - Services generated by langium-cli
 *  - Services specified in this file
 *
 * @param context Optional module context with the LSP connection.
 * @param options Further options to configure the Safe-DS module.
 * @return An object wrapping the shared services and the language-specific services.
 */
export const createSafeDsServicesWithBuiltins = async function (
    context: DefaultSharedModuleContext,
    options?: ModuleOptions,
): Promise<{
    shared: LangiumSharedServices;
    SafeDs: SafeDsServices;
}> {
    const { shared, SafeDs } = createSafeDsServices(context, options);
    await shared.workspace.WorkspaceManager.initializeWorkspace([]);
    return { shared, SafeDs };
};

/**
 * Options to pass to the creation of Safe-DS services.
 */
export interface ModuleOptions {
    /**
     * Optional command to start the runner, if available.
     */
    runnerCommand?: string;
}
