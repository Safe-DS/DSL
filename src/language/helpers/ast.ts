import { AstNode, hasContainerOfType } from 'langium';

/**
 * Returns whether the inner node is contained in the outer node. If the nodes are equal, this function returns `true`.
 */
export const isContainedIn = (inner: AstNode | undefined, outer: AstNode | undefined): boolean => {
    return hasContainerOfType(inner, (it) => it === outer);
};
