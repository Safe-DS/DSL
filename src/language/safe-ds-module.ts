import {
    createDefaultModule,
    createDefaultSharedModule,
    DeepPartial,
    DefaultSharedModuleContext,
    inject,
    LangiumServices,
    LangiumSharedServices,
    Module,
    PartialLangiumServices,
} from 'langium';
import { SafeDsGeneratedModule, SafeDsGeneratedSharedModule } from './generated/module.js';
import { registerValidationChecks, SafeDsValidator } from './validation/safe-ds-validator.js';
import { SafeDsFormatter } from './formatting/safe-ds-formatter.js';
import { SafeDsWorkspaceManager } from './builtins/safe-ds-workspace-manager.js';
import { SafeDsScopeComputation } from './scoping/safe-ds-scope-computation.js';
import { SafeDsScopeProvider } from './scoping/safe-ds-scope-provider.js';

/**
 * Declaration of custom services - add your own service classes here.
 */
export type SafeDsAddedServices = {
    validation: {
        SafeDsValidator: SafeDsValidator;
    };
};

/**
 * Union of Langium default services and your custom services - use this as constructor parameter
 * of custom service classes.
 */
export type SafeDsServices = LangiumServices & SafeDsAddedServices;

/**
 * Dependency injection module that overrides Langium default services and contributes the
 * declared custom services. The Langium defaults can be partially specified to override only
 * selected services, while the custom services must be fully specified.
 */
export const SafeDsModule: Module<SafeDsServices, PartialLangiumServices & SafeDsAddedServices> = {
    lsp: {
        Formatter: () => new SafeDsFormatter(),
    },
    references: {
        ScopeComputation: (services) => new SafeDsScopeComputation(services),
        ScopeProvider: (services) => new SafeDsScopeProvider(services),
    },
    validation: {
        SafeDsValidator: () => new SafeDsValidator(),
    },
};

export type SafeDsSharedServices = LangiumSharedServices;

export const SafeDsSharedModule: Module<SafeDsSharedServices, DeepPartial<SafeDsSharedServices>> = {
    workspace: {
        WorkspaceManager: (services) => new SafeDsWorkspaceManager(services),
    },
};

/**
 * Create the full set of services required by Langium.
 *
 * First inject the shared services by merging two modules:
 *  - Langium default shared services
 *  - Services generated by langium-cli
 *
 * Then inject the language-specific services by merging three modules:
 *  - Langium default language-specific services
 *  - Services generated by langium-cli
 *  - Services specified in this file
 *
 * @param context Optional module context with the LSP connection
 * @return An object wrapping the shared services and the language-specific services
 */
export const createSafeDsServices = function (context: DefaultSharedModuleContext): {
    shared: LangiumSharedServices;
    SafeDs: SafeDsServices;
} {
    const shared = inject(createDefaultSharedModule(context), SafeDsGeneratedSharedModule, SafeDsSharedModule);
    const SafeDs = inject(createDefaultModule({ shared }), SafeDsGeneratedModule, SafeDsModule);
    shared.ServiceRegistry.register(SafeDs);
    registerValidationChecks(SafeDs);
    return { shared, SafeDs };
};
