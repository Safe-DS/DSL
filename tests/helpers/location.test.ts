import { describe, expect, it } from 'vitest';
import { getNodeByLocation, isLocationEqual, locationToString, positionToString, rangeToString } from './location.js';
import { createSafeDsServices } from '../../src/language/safe-ds-module.js';
import { EmptyFileSystem } from 'langium';
import { AssertionError } from 'assert';
import { parseHelper } from 'langium/test';
import { isSdsClass } from '../../src/language/generated/ast.js';

describe('positionToString', () => {
    it.each([
        {
            position: { line: 0, character: 0 },
            expected: '1:1',
        },
        {
            position: { line: 1, character: 0 },
            expected: '2:1',
        },
    ])('should convert position to string ($expected)', ({ position, expected }) => {
        expect(positionToString(position)).toBe(expected);
    });
});

describe('rangeToString', () => {
    it.each([
        {
            range: { start: { line: 0, character: 0 }, end: { line: 0, character: 0 } },
            expected: '1:1 -> 1:1',
        },
        {
            range: { start: { line: 0, character: 0 }, end: { line: 1, character: 0 } },
            expected: '1:1 -> 2:1',
        },
    ])('should convert range to string ($expected)', ({ range, expected }) => {
        expect(rangeToString(range)).toBe(expected);
    });
});

describe('locationToString', () => {
    it.each([
        {
            location: {
                uri: 'file:///test.sdstest',
                range: { start: { line: 0, character: 0 }, end: { line: 0, character: 0 } },
            },
            expected: 'file:///test.sdstest:1:1 -> 1:1',
        },
        {
            location: {
                uri: 'file:///test.sdstest',
                range: { start: { line: 0, character: 0 }, end: { line: 1, character: 0 } },
            },
            expected: 'file:///test.sdstest:1:1 -> 2:1',
        },
    ])(`should convert location to string ($expected)`, ({ location, expected }) => {
        expect(locationToString(location)).toBe(expected);
    });
});

describe('isLocationEqual', () => {
    it.each([
        {
            location1: {
                uri: 'file:///test.sdstest',
                range: { start: { line: 0, character: 0 }, end: { line: 0, character: 0 } },
            },
            location2: {
                uri: 'file:///test.sdstest',
                range: { start: { line: 0, character: 0 }, end: { line: 0, character: 0 } },
            },
            expected: true,
            id: 'same location',
        },
        {
            location1: {
                uri: 'file:///test.sdstest',
                range: { start: { line: 0, character: 0 }, end: { line: 0, character: 0 } },
            },
            location2: {
                uri: 'file:///test2.sdstest',
                range: { start: { line: 0, character: 0 }, end: { line: 0, character: 0 } },
            },
            expected: false,
            id: 'different uri',
        },
        {
            location1: {
                uri: 'file:///test.sdstest',
                range: { start: { line: 0, character: 0 }, end: { line: 0, character: 0 } },
            },
            location2: {
                uri: 'file:///test.sdstest',
                range: { start: { line: 0, character: 0 }, end: { line: 1, character: 0 } },
            },
            expected: false,
            id: 'different range',
        },
    ])('should compare locations for equality ($id)', ({ location1, location2, expected }) => {
        expect(isLocationEqual(location1, location2)).toBe(expected);
    });
});

describe('getNodeByLocation', () => {
    it('should raise if no document is found', () => {
        const services = createSafeDsServices(EmptyFileSystem).SafeDs;

        expect(() => {
            getNodeByLocation(services, {
                uri: 'file:///test.sdstest',
                range: { start: { line: 0, character: 0 }, end: { line: 0, character: 0 } },
            });
        }).toThrowErrorMatchingSnapshot();
    });

    it('should raise if no node is found', async () => {
        const services = createSafeDsServices(EmptyFileSystem).SafeDs;
        const document = await parseHelper(services)(`class C`);

        expect(() => {
            getNodeByLocation(services, {
                uri: document.uri.toString(),
                range: { start: { line: 0, character: 0 }, end: { line: 0, character: 0 } },
            });
        }).toThrow(AssertionError);
    });

    it('should return the node that fills the range completely', async () => {
        const services = createSafeDsServices(EmptyFileSystem).SafeDs;
        const document = await parseHelper(services)(`class C`);

        expect(
            getNodeByLocation(services, {
                uri: document.uri.toString(),
                range: { start: { line: 0, character: 0 }, end: { line: 0, character: 7 } },
            }),
        ).to.satisfy(isSdsClass);
    });

    it('should return the node whose name fills the range completely', async () => {
        const services = createSafeDsServices(EmptyFileSystem).SafeDs;
        const document = await parseHelper(services)(`class C`);

        expect(
            getNodeByLocation(services, {
                uri: document.uri.toString(),
                range: { start: { line: 0, character: 6 }, end: { line: 0, character: 7 } },
            }),
        ).to.satisfy(isSdsClass);
    });
});
