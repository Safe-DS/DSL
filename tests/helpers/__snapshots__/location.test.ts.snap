// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getNodeByLocation > should raise if no document is found 1`] = `"No document found at file:///test.sdstest"`;
